### PR TITLE
Add `useInternetConnection` hook to track internet connectivity changes

### PR DESCRIPTION
**Pull request by romin-halltari**
### 📦 Pull Request

[Provide a general summary of the pull request here.]

### ✅ Fixed Issues

- [List any fixed issues here like: Fixes #XXXX]

### 🚨 Test instructions

[Describe any additional context required to test the PR/feature/bug fix.]

### ⚠️ Don't forget to add a [semver](https://semver.org/) label! 
Please only add **one** label:
- `patch`: Bug Fix?
- `minor`: New Feature?
- `major`: Breaking Change?
- `skip-release`: It's unnecessary to publish this change.

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @magic-ext/algorand@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/aptos@4.3.0-canary.665.6934462457.0
  npm install @magic-ext/auth@4.3.0-canary.665.6934462457.0
  npm install @magic-ext/avalanche@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/bitcoin@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/conflux@14.3.0-canary.665.6934462457.0
  npm install @magic-ext/cosmos@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/ed25519@12.3.0-canary.665.6934462457.0
  npm install @magic-ext/flow@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/gdkms@4.3.0-canary.665.6934462457.0
  npm install @magic-ext/harmony@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/icon@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/near@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/oauth@15.3.0-canary.665.6934462457.0
  npm install @magic-ext/oidc@4.3.0-canary.665.6934462457.0
  npm install @magic-ext/polkadot@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/react-native-bare-oauth@17.1.0-canary.665.6934462457.0
  npm install @magic-ext/react-native-expo-oauth@17.1.0-canary.665.6934462457.0
  npm install @magic-ext/solana@18.1.0-canary.665.6934462457.0
  npm install @magic-ext/taquito@13.3.0-canary.665.6934462457.0
  npm install @magic-ext/terra@13.3.0-canary.665.6934462457.0
  npm install @magic-ext/tezos@16.3.0-canary.665.6934462457.0
  npm install @magic-ext/webauthn@15.3.0-canary.665.6934462457.0
  npm install @magic-ext/zilliqa@16.3.0-canary.665.6934462457.0
  npm install @magic-sdk/commons@17.3.0-canary.665.6934462457.0
  npm install @magic-sdk/pnp@15.3.0-canary.665.6934462457.0
  npm install @magic-sdk/provider@21.3.0-canary.665.6934462457.0
  npm install @magic-sdk/react-native-bare@22.3.0-canary.665.6934462457.0
  npm install @magic-sdk/react-native-expo@22.3.0-canary.665.6934462457.0
  npm install magic-sdk@21.3.0-canary.665.6934462457.0
  # or 
  yarn add @magic-ext/algorand@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/aptos@4.3.0-canary.665.6934462457.0
  yarn add @magic-ext/auth@4.3.0-canary.665.6934462457.0
  yarn add @magic-ext/avalanche@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/bitcoin@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/conflux@14.3.0-canary.665.6934462457.0
  yarn add @magic-ext/cosmos@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/ed25519@12.3.0-canary.665.6934462457.0
  yarn add @magic-ext/flow@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/gdkms@4.3.0-canary.665.6934462457.0
  yarn add @magic-ext/harmony@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/icon@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/near@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/oauth@15.3.0-canary.665.6934462457.0
  yarn add @magic-ext/oidc@4.3.0-canary.665.6934462457.0
  yarn add @magic-ext/polkadot@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/react-native-bare-oauth@17.1.0-canary.665.6934462457.0
  yarn add @magic-ext/react-native-expo-oauth@17.1.0-canary.665.6934462457.0
  yarn add @magic-ext/solana@18.1.0-canary.665.6934462457.0
  yarn add @magic-ext/taquito@13.3.0-canary.665.6934462457.0
  yarn add @magic-ext/terra@13.3.0-canary.665.6934462457.0
  yarn add @magic-ext/tezos@16.3.0-canary.665.6934462457.0
  yarn add @magic-ext/webauthn@15.3.0-canary.665.6934462457.0
  yarn add @magic-ext/zilliqa@16.3.0-canary.665.6934462457.0
  yarn add @magic-sdk/commons@17.3.0-canary.665.6934462457.0
  yarn add @magic-sdk/pnp@15.3.0-canary.665.6934462457.0
  yarn add @magic-sdk/provider@21.3.0-canary.665.6934462457.0
  yarn add @magic-sdk/react-native-bare@22.3.0-canary.665.6934462457.0
  yarn add @magic-sdk/react-native-expo@22.3.0-canary.665.6934462457.0
  yarn add magic-sdk@21.3.0-canary.665.6934462457.0
  ```
</details>

### CODE REVIEW

File: packages/@magic-sdk/react-native-bare/test/mocks.ts [hunk 1]
```diff
@@ -6,10 +44,9 @@ export function reactNativeStyleSheetStub() {
 const noopModule = () => ({});
 
 export function removeReactDependencies() {
-  jest.mock('react', noopModule);
```
Comment (Ariflo) [inline packages/@magic-sdk/react-native-bare/test/mocks.ts:9]: Is there a reason you chose to remove this line? In other words, is there a reason we'd previously removed the react mock on setup?

File: packages/@magic-sdk/react-native-bare/test/spec/hooks.spec.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  it('should call the listener when the connection changes', async () => {
+    NetInfo.addEventListener = jest.fn();
+
+    const { result } = renderHook(() => useInternetConnection());
+
+    // Initial render, assuming it's connected
+    expect(result.current).toBe(true);
+
+    // Simulate a change in connection status
+    act(() => {
```
Comment (Ariflo) [inline packages/@magic-sdk/react-native-bare/test/spec/hooks.spec.ts:34]: This is cool :sunglasses:

File: packages/@magic-sdk/react-native-bare/src/index.ts [hunk 3]
```diff
@@ -69,3 +69,5 @@ export type Magic<T extends MagicSDKExtensionsOption<any> = MagicSDKExtensionsOp
   SDKBaseReactNative,
   T
 >;
+
+export { useInternetConnection } from './hooks';
```
Comment (Ariflo) [inline packages/@magic-sdk/react-native-bare/src/index.ts:73]: This is close, but not exactly what I believe we were hoping for to resolve this issue. When we spoke, I was under the impression that the idea was that we'd utilize something like this within the SDK and return an error to the developer. From what I can tell [here](https://github.com/magiclabs/magic-js/pull/665/files#diff-c7d7b729e018c26c53125262037c912483ac4ac2f49171243549d87035c8040cR88-R113), the developer would need to add boiler plate around the `<m.relayer />` to detect the internet failure. 

Have you considered potentially adding the check for internet before [this invocation](https://github.com/magiclabs/magic-js/blob/e7ab328946259baa518e43ddf4e010b11df1da08/packages/%40magic-sdk/react-native-expo/src/index.ts#L47-L66)?  Will RN spit out a readable error if a component like `<m.relayer />` `throw`s an error? LMK your thoughts here.

**Review by Ariflo — CHANGES_REQUESTED** [review]
Left some comments on the overall approach here. LMK 👍

File: packages/@magic-sdk/react-native-bare/test/mocks.ts (hunk 1 above)
Comment (romin-halltari) [reply packages/@magic-sdk/react-native-bare/test/mocks.ts:9]: I'm not sure why it was mocked, normally in RN projects you don't mock all of React. Also, removing this line doesn't cause any issues with the tests. Leaving it as it was does cause issues when testing things like hooks that come from React (useState, useEffect, etc.).

**Review by Ariflo — CHANGES_REQUESTED** [review]

File: packages/@magic-sdk/provider/src/core/view-controller.ts [hunk 4]
```diff
@@ -88,6 +90,7 @@ async function persistMagicEventRefreshToken(event: MagicMessageEvent) {
 }
 
 export abstract class ViewController {
+  private isReady = false;
```
Comment (Ariflo) [inline packages/@magic-sdk/provider/src/core/view-controller.ts:93]: nit: The naming convention here is little confusing with `this.ready` already here. Is there a better name to describe what this is actually checking for?

File: packages/@magic-sdk/react-native-bare/README.md [hunk 5]
```diff
@@ -69,4 +71,44 @@ Please note that as of **v14.0.0** our React Native package offerings wrap the `
...
+
+<img width="299" alt="Screenshot 2022-11-23 at 12 19 19 PM" src="https://user-images.githubusercontent.com/13407884/203641477-ec2e472e-86dc-4a22-b54a-eb694001617e.png">
+
+When attempting to import `Magic`, take note that the React Native metro bundler doesn’t work well with symlinks, which tend to be utilized by most package managers.
+
+For this issue consider using Microsoft's [rnx-kit](https://microsoft.github.io/rnx-kit/docs/guides/bundling) suite of tools that include a plugin for metro that fixes this symlink related error.
+
+### Handling internet connection problems
+
+When the app has internet connectivity issues, the relayer might behave in ways you don't expect. That is why it is important to use [@react-native-community/netinfo](https://www.npmjs.com/package/@react-native-community/netinfo) to track the internet connection state of the device, and refresh your UI when the connection is re-established. For your convenience, we've also added a hook that uses this library behind the scenes:
```
Comment (Ariflo) [inline packages/@magic-sdk/react-native-bare/README.md:90]: Any feelings on calling out the new error message that will be returned here?

**Review by Ariflo — CHANGES_REQUESTED** [review]
Just a few small comments. Looking good!

File: packages/@magic-sdk/provider/src/core/view-controller.ts (hunk 4 above)
Comment (Ariflo) [reply packages/@magic-sdk/provider/src/core/view-controller.ts:93]: I like it, go for it!

File: packages/@magic-sdk/react-native-bare/README.md (hunk 5 above)
Comment (romin-halltari) [reply packages/@magic-sdk/react-native-bare/README.md]: Updated 👍

**Review by Ariflo — APPROVED** [review]
LGTM, thanks for all the updates!

**Comment by Ariflo** [comment]
:rocket: PR was released in [`@magic-ext/algorand@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/algorand%4016.3.0), [`@magic-ext/aptos@4.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/aptos%404.3.0), [`@magic-ext/auth@4.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/auth%404.3.0), [`@magic-ext/avalanche@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/avalanche%4016.3.0), [`@magic-ext/bitcoin@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/bitcoin%4016.3.0), [`@magic-ext/conflux@14.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/conflux%4014.3.0), [`@magic-ext/cosmos@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/cosmos%4016.3.0), [`@magic-ext/ed25519@12.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/ed25519%4012.3.0), [`@magic-ext/flow@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/flow%4016.3.0), [`@magic-ext/gdkms@4.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/gdkms%404.3.0), [`@magic-ext/harmony@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/harmony%4016.3.0), [`@magic-ext/icon@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/icon%4016.3.0), [`@magic-ext/near@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/near%4016.3.0), [`@magic-ext/oauth@15.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/oauth%4015.3.0), [`@magic-ext/oidc@4.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/oidc%404.3.0), [`@magic-ext/polkadot@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/polkadot%4016.3.0), [`@magic-ext/react-native-bare-oauth@17.1.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/react-native-bare-oauth%4017.1.0), [`@magic-ext/react-native-expo-oauth@17.1.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/react-native-expo-oauth%4017.1.0), [`@magic-ext/solana@18.1.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/solana%4018.1.0), [`@magic-ext/taquito@13.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/taquito%4013.3.0), [`@magic-ext/terra@13.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/terra%4013.3.0), [`@magic-ext/tezos@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/tezos%4016.3.0), [`@magic-ext/webauthn@15.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/webauthn%4015.3.0), [`@magic-ext/zilliqa@16.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-ext/zilliqa%4016.3.0), [`@magic-sdk/commons@17.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-sdk/commons%4017.3.0), [`@magic-sdk/pnp@15.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-sdk/pnp%4015.3.0), [`@magic-sdk/provider@21.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-sdk/provider%4021.3.0), [`@magic-sdk/react-native-bare@22.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-sdk/react-native-bare%4022.3.0), [`@magic-sdk/react-native-expo@22.3.0`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-sdk/react-native-expo%4022.3.0), [`magic-sdk@21.3.0`](https://github.com/magiclabs/magic-js/releases/tag/magic-sdk%4021.3.0) :rocket:

**Pull request merged**